### PR TITLE
Add Api function to get random tag value add key authentication

### LINKED ISSUE #4: There should be an API to perform searches on records by a tag name

### LINKED ISSUE #6: There should be an environmental variable API key to interface with the API

### PR DESCRIPTION
**Pull request by brentm5**
Closes #4 & Closes #6

### CODE REVIEW

File: app/controllers/api/base_api_controller.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class Api::BaseApiController < ActionController::Base
+  before_filter :authenticate_user
+
+  private
+  def authenticate_user
+    if api_token_blank || invalid_api_token
```
Comment (drapergeek) [inline app/controllers/api/base_api_controller.rb]: Does it matter if the api token is blank?  All we care is if its valid or not? If they send a blank token then it won't be valid.

File: app/controllers/api/base_api_controller.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      render json: { error: 'Token is invalid' }, status: 401
+    end
+  end
+
+  def api_token_blank
+    ENV['API_TOKEN'] == nil || ENV['API_TOKEN'] == ''
```
Comment (drapergeek) [inline app/controllers/api/base_api_controller.rb]: You are replicating the functionality of the `blank?` method. You could use that instead.

File: app/controllers/api/random_mapped_value_controller.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+class Api::RandomMappedValueController < Api::BaseApiController
+  before_filter :check_parameters
+
+  def index
+    tag = Tag.find_by_name(params[:tag])
```
Comment (drapergeek) [inline app/controllers/api/random_mapped_value_controller.rb]: You could move tag to a private method that gets lazily evaluated.

```
private

def tag
 @tag ||= Tag.find_by_name(params[:tag])
end

```

File: app/controllers/api/random_mapped_value_controller.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    if tag
+      values = Tag.find_by_name(params[:tag]).mapped_values
```
Comment (drapergeek) [inline app/controllers/api/random_mapped_value_controller.rb]: Why are you looking up the tag again if you're already found it once?

File: app/controllers/api/random_mapped_value_controller.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      if values.length > 0
```
Comment (drapergeek) [inline app/controllers/api/random_mapped_value_controller.rb]: I would also move the values to another method that is named better.

File: app/controllers/api/random_mapped_value_controller.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        render json: values.sample
+        return
+      end
+    end
+    render json: {}
+  end
+
+  private 
+  def check_parameters
```
Comment (drapergeek) [inline app/controllers/api/random_mapped_value_controller.rb]: Check parameters is very ambiguous.  How about `ensure_tag_parameter`?

File: spec/api/mapped_value_spec.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+describe Api::RandomMappedValueController, type: :controller do
+  let(:token) { ENV['API_TOKEN'] }
+
+  context 'Successfuly requests' do
+    it 'can get a random mapped value with a tag' do
+      mapped_value = create :mapped_value
+      get :index, token: token, tag: mapped_value.tag.name
+      response.body.should eql mapped_value.to_json
+      response.status.should eql 200
+    end
```
Comment (drapergeek) [inline spec/api/mapped_value_spec.rb:14]: White space.

File: spec/api/mapped_value_spec.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    it 'should response with an empty object if the tag has no values associated' do
+      tag = create :tag
+      get :index, token: token, tag: tag.name
+      response.body.should eql '{}'
+      response.status.should eql 200
+    end 
+  end
+
+  context 'Invalid Requests' do
+    it 'do not pass through a token' do
```
Comment (drapergeek) [inline spec/api/mapped_value_spec.rb]: What does 'it does not pass through a token` even mean?

File: spec/api/mapped_value_spec.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+      mapped_value = create :mapped_value
+      get :index, tag: mapped_value.tag.name
+      response.status.should eql 401
+    end
+    it 'do not pass through a tag' do
```
Comment (drapergeek) [inline spec/api/mapped_value_spec.rb]: See above.

File: spec/api/mapped_value_spec.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      get :index, token: token, tag: mapped_value.tag.name
+      response.body.should eql mapped_value.to_json
+      response.status.should eql 200
+    end
+    it 'should respond with an empty object if not tags are found' do
+      get :index, token: token, tag: 'this-does-not-exist'
+      response.body.should eql '{}'
+      response.status.should eql 200
+    end
+    it 'should response with an empty object if the tag has no values associated' do
```
Comment (drapergeek) [inline spec/api/mapped_value_spec.rb]: "It should RESPOND"

File: app/controllers/api/base_api_controller.rb (hunk 1 above)
Comment (brentm5) [inline app/controllers/api/base_api_controller.rb]: This has to do with the one set.  If there isnt one sent then noneone should be able to access it

File: app/controllers/api/random_mapped_value_controller.rb (hunk 4 above)
Comment (brentm5) [inline app/controllers/api/random_mapped_value_controller.rb]: this is what i would like to call a simple overlooking... as that honestly makes no sense :-1:

File: spec/api/mapped_value_spec.rb (hunk 10 above)
Comment (brentm5) [inline spec/api/mapped_value_spec.rb]: pish posh

File: app/controllers/api/base_api_controller.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
+class Api::BaseApiController < ActionController::Base
+  before_filter :authenticate_user
+
+  private
+  def authenticate_user
+    if api_token_not_set || invalid_api_token
```
Comment (drapergeek) [inline app/controllers/api/base_api_controller.rb:6]: I still don't know why you even care if its blank? Why not just make sure it doesn't match?

File: app/controllers/api/random_mapped_value_controller.rb [hunk 12]
```diff
@@ -0,0 +1,28 @@
+class Api::RandomMappedValueController < Api::BaseApiController
+  before_filter :ensure_tag_parameter
+
+  def index
```
Comment (drapergeek) [inline app/controllers/api/random_mapped_value_controller.rb:4]: This is so much cleaner!

File: spec/api/mapped_value_spec.rb [hunk 13]
```diff
@@ -0,0 +1,51 @@
+require 'spec_helper'
+
+describe Api::RandomMappedValueController, type: :controller do
+  let(:token) { ENV['API_TOKEN'] }
```
Comment (drapergeek) [inline spec/api/mapped_value_spec.rb:4]: You really really reallly really really should avoid lets.

http://robots.thoughtbot.com/post/34709581001/lets-not

Trust me, I'm working on stuff right now and the lets are causing a LOT of issues.

Comment (brentm5) [inline spec/api/mapped_value_spec.rb:4]: Ill have you know i removed the lets statements when adding new features.  Took me forever to find where you posted this haha

**Pull request closed without merging**